Newsletters: add 280 (2023-12-06) (#1421)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2023-12-06-newsletter.md

---
title: 'Bitcoin Optech Newsletter #280'
permalink: /en/newsletters/2023/12/06/
name: 2023-12-06-newsletter
slug: 2023-12-06-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter FIXME

## News

- **Cluster mempool discussion:** Bitcoin Core developers working on
  [cluster mempool][topic cluster mempool] started a [working
  group][wg-cluster-mempool] (WG) on the Delving Bitcoin forum. Cluster
  mempool is a proposal to make it much easier to compare the
  profitability of different valid orderings of transactions.  A valid
  ordering of Bitcoin transactions requires parent transactions get
  confirmed before their children, either by putting the parent in an
  earlier block than any of its children, or earlier in the same block.
  _Clusters_ of one or more related transactions are designed to be easy
  to include into feerate-sorted _chunks_ that are each independent of

>   the other chunks. For example, we might split a mempool into 100

Reviewer: I think there might be misunderstanding here. A _cluster_ refers to the connected component of a transaction, i.e. all ancestors and descendants of a transaction all of the ancestors and descendants of all those, plus all of the ancestors and descendants of all that, etc. until there is nothing to add. _Chunks_ refer to sections of a linearization of a cluster. 

Linearizations must be topologically valid. They are chunked by first making every transaction its own chunk, then combining any two chunks A and B, if A is the left neighbor of B and B has a higher feerate than A. Ideally, linearizations are sorted to produce the highest possible feerate diagram across the cluster, but non-optimal linearizations may occur.

We are currently expecting that clusters might be limited to 100 transactions, much like the descendant and ancestor limit in the current mempool design, and since chunks are subcomponents of a cluster’s linearization, they would usually not be whole blocks.

>   chunks each about 4 million weight in size, each of which could be

Reviewer: Having calculated a linearization for every cluster in our mempool and subdivided it into chunks, we can now create a total order of all chunks in our mempool. We only need to compare the best unselected chunks in each cluster, rather than each transaction with its ancestry. By picking the best unselected chunk from a cluster repeatedly, we can easily assemble a block template, and vice versa, if we evict, we can simply compare the worst feerate chunks from each cluster to see which one we would pick into a block last.

Reviewer 2: @murchandamus Note that to assemble the block we're pick the best chunks from *all* clusters in the mempool, simply comparing them by feerate (and perhaps size).

@sipa It makes sense to me that chunks need to be limited in size (as we add chunks to a block as individual units), but not why clusters necessarily need to have the same limit. I think the cluster limit results from the possibility that the optimal chunking of a cluster may end up being almost the entire cluster (particularly when multiple children pay for a parent, eg), and if we didn't limit the cluster size, we could end up stuck with a suboptimal linearisation, which we don't want. Maybe some commentary about limits would make for a good post as well?

  mined successively into a block.

>     By comparing different lists of chunks, we can determine which would

Reviewer: "Chunkings across multiple clusters" would only occur when we are trying to determine whether a replacement is strictly better than our current mempool content. E.g. a replacement transaction could combine two previously unconnected clusters, and then we’d have to create combined feerate diagram across both these clusters to compare the state with the two separate clusters to the state after the replacement.

    - [Merging incomparable linearizations][cluster merge] looks at how
      to merge two different sets of chunks ("chunkings") for the same
      set of transactions, specifically _incomparable_ chunkings.
      Chunkings can be compared if one of them always accumulates the
      same or greater amount of fees within any number of vbytes
      (discrete to the size of the chunks).  For example:

      ![Comparable chunkings](/img/posts/2023-12-comparable-chunkings.png)

      Chunkings are incomparable if one of them accumulates a greater
      amount of fees within some number of vbytes but the other chunking
      accumulates a greater amount of fees within a larger number of
      vbytes.  For example:

      ![Incomparable chunkings](/img/posts/2023-12-incomparable-chunkings.png)

      As one of the theorems in the previously linked thread notes, "if
      one has two incomparable chunkings for a graph, then another
      chunking must exist which is strictly better than both".  That
      means an effective method for merging two different incomparable
      chunkings can be a powerful tool for improving miner

>       profitability.  For example, miners Alice and Bob could both try

Reviewer: Not just miners, but any node could try to find better linearizations

      same or greater amount of fees within any number of vbytes
      (discrete to the size of the chunks).  For example:

      ![Comparable chunkings](/img/posts/2023-12-comparable-chunkings.png)

      Chunkings are incomparable if one of them accumulates a greater
      amount of fees within some number of vbytes but the other chunking
      accumulates a greater amount of fees within a larger number of
      vbytes.  For example:

      ![Incomparable chunkings](/img/posts/2023-12-incomparable-chunkings.png)

      As one of the theorems in the previously linked thread notes, "if
      one has two incomparable chunkings for a graph, then another
      chunking must exist which is strictly better than both".  That
      means an effective method for merging two different incomparable
      chunkings can be a powerful tool for improving miner
      profitability.  For example, miners Alice and Bob could both try
      to find the optimal chunking and could share their results with
      each other in a way that would guarantee they'd both obtain a
      third chunking that was at least as good as either of their
      original chunkings---and possibly better.

>     - [Post-cluster package RBF][cluster rbf] discusses an alternative

Reviewer 3: you might want to note we don't even *have* package rbf today, and what is proposed pre-cluster mempool is very limited

      to the rules currently used for [replace by fee][topic rbf], which
      are known to cause multiple problems, including [pinning
      attacks][topic transaction pinning].  The goal is to accept any
      replacement transaction or [package of transactions][topic package
      relay] that will produce a chunking that is better than an
      equivalent chunking of the current mempool (as long as the
      replacement also follows normal anti-DoS relay policy, such as
      having the replacement pay for its own relay by the _incremental
      relay feerate_, which is currently 1 sat per vbyte).  Several
      [other][cluster rbf-old1] threads [also][cluster rbf-old2]
      discussed [this][cluster rbf-old3] topic.

- **Testing with warnet:** Matthew Zipkin [posted][zipkin warnet] to
  Delving Bitcoin with the results of some simulations he's run using
  [warnet][], a program that launches a large number of Bitcoin nodes
  with a defined set of connections between them (usually on a test

>   network).  Zipkin results show the extra memory that would be used if

Reviewer: ```suggestion
  network).  Zipkin's results show the extra memory that would be used if
```

  several proposed changes to peer management code are merged (either
  independently or together).  He also notes that he's excited to use
  simulations for testing other proposed changes and to quantify the
  effect of proposed attacks.

## Bitcoin Core PR Review Club

*In this monthly section, we summarize a recent [Bitcoin Core PR Review
Club][] meeting, highlighting some of the important questions and
answers.  Click on a question below to see a summary of the answer from
the meeting.*

FIXME:LarryRuane

{% include functions/details-list.md
  q0="FIXME"
  a0="FIXME"
  a0link="https://bitcoincore.reviews/28368#l-30FIXME"
%}

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

> - [Bitcoin Core 26.0rc3][] is a release candidate for the next major

Reviewer: rc4 and final got tagged meanwhile

>   version of the predominant full-node implementation. There's a [testing

Reviewer 4: Per https://github.com/bitcoinops/bitcoinops.github.io/pull/1400#discussion_r1402368137:
```suggestion
  version of the predominant full node implementation. There's a [testing
```

Reviewer 5: It reads better to me with the hyphen, because it's conceptually a single, unitary thing (rather than saying it's a node that happens to be "full" whatever that means -- is it out of disk space?), but I don't have a strong opinion.

Reviewer 4: I'm not knowledgeble enough to discuss the language perspective, I'm just saying this goes against long established terminology both at Bitcoin Optech and most everywhere else I could find. The hyphenated version was never used on Optech until a few weeks ago. I'd think we'd like to keep terminology consistent since we even have a style guide for this very purpose.

Author: @LarryRuane would you hyphenate "Bitcoin node implementation" and "LN node implementation"?

Given that Optech frequently discusses things like peer networks that can be thought of topologically as represented by nodes and edges, we modify the noun "node" a lot to make it clear what type of node we're talking about, so we probably want a general policy.

---
title: 'Bitcoin Optech Newsletter #280'
permalink: /en/newsletters/2023/12/06/
name: 2023-12-06-newsletter
slug: 2023-12-06-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter FIXME

## News

- **Cluster mempool discussion:** Bitcoin Core developers working on
  [cluster mempool][topic cluster mempool] started a [working
  group][wg-cluster-mempool] (WG) on the Delving Bitcoin forum. Cluster
  mempool is a proposal to make it much easier to compare the
  profitability of different valid orderings of transactions.  A valid

>   ordering of Bitcoin transactions requires parent transactions get

Reviewer 5: ```suggestion
  ordering of Bitcoin transactions requires parent transactions to be
```

  confirmed before their children, either by putting the parent in an
  earlier block than any of its children, or earlier in the same block.
  _Clusters_ of one or more related transactions are designed to be easy
  to include into feerate-sorted _chunks_ that are each independent of
  the other chunks. For example, we might split a mempool into 100
  chunks each about 4 million weight in size, each of which could be
  mined successively into a block.

    By comparing different lists of chunks, we can determine which would
    be better for miners.  Any change that increases the total fees of
    the first chunk (in the same number of vbytes or less) is an

>     improvement to a miner; any change that does the opposite (reduce

Reviewer 5: ```suggestion
    improvement to a miner; any change that does the opposite (reduces
```

      to merge two different sets of chunks ("chunkings") for the same
      set of transactions, specifically _incomparable_ chunkings.
      Chunkings can be compared if one of them always accumulates the
      same or greater amount of fees within any number of vbytes
      (discrete to the size of the chunks).  For example:

      ![Comparable chunkings](/img/posts/2023-12-comparable-chunkings.png)

      Chunkings are incomparable if one of them accumulates a greater
      amount of fees within some number of vbytes but the other chunking
      accumulates a greater amount of fees within a larger number of
      vbytes.  For example:

      ![Incomparable chunkings](/img/posts/2023-12-incomparable-chunkings.png)

      As one of the theorems in the previously linked thread notes, "if
      one has two incomparable chunkings for a graph, then another
      chunking must exist which is strictly better than both".  That
      means an effective method for merging two different incomparable
      chunkings can be a powerful tool for improving miner
      profitability.  For example, miners Alice and Bob could both try

>       to find the optimal chunking and could share their results with

Reviewer 5: ```suggestion
      to find the optimal chunking and share their results with
```

  a4link="https://bitcoincore.reviews/v26-rc-testing#l-115"
%}

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

- [Bitcoin Core 26.0rc3][] is a release candidate for the next major
  version of the predominant full-node implementation. There's a [testing
  guide][26.0 testing] available.

<!-- FIXME:harding to update Tuesday -->

## Notable code and documentation changes

*Notable recent changes in [Bitcoin Core][bitcoin core repo], [Core
Lightning][core lightning repo], [Eclair][eclair repo], [LDK][ldk repo],
[LND][lnd repo], [libsecp256k1][libsecp256k1 repo], [Hardware Wallet
Interface (HWI)][hwi repo], [Rust Bitcoin][rust bitcoin repo], [BTCPay
Server][btcpay server repo], [BDK][bdk repo], [Bitcoin Improvement
Proposals (BIPs)][bips repo], [Lightning BOLTs][bolts repo], and
[Bitcoin Inquisition][bitcoin inquisition repo].*

- [Bitcoin Core #28848][] updates the `submitpackage` RPC to be more helpful when
  any transaction fails. Instead of throwing a JSONRPCError with the first

>   failure, it return results for each transaction whenever possible.

Reviewer 5: ```suggestion
  failure, it returns results for each transaction whenever possible.
```

        Core to determine whether a transaction should be replaced,
        potentially improving the RBF rules in several areas, including
        proposed [package relay][topic package relay] for replacements.
        Several [other][cluster rbf-old1] threads [also][cluster
        rbf-old2] discussed [this][cluster rbf-old3] topic.

- **Testing with warnet:** Matthew Zipkin [posted][zipkin warnet] to
  Delving Bitcoin with the results of some simulations he's run using
  [warnet][], a program that launches a large number of Bitcoin nodes
  with a defined set of connections between them (usually on a test
  network).  Zipkin results show the extra memory that would be used if
  several proposed changes to peer management code are merged (either
  independently or together).  He also notes that he's excited to use
  simulations for testing other proposed changes and to quantify the
  effect of proposed attacks.

## Bitcoin Core PR Review Club

*In this monthly section, we summarize a recent [Bitcoin Core PR Review
Club][] meeting, highlighting some of the important questions and
answers.  Click on a question below to see a summary of the answer from
the meeting.*

> Before each [major Bitcoin Core release][], extensive testing by the

Reviewer 6: I think the summary for this PR Review Club is great and we could likely drop all of the Q&A section below, which I thought added less value.

Reviewer 5: > I think the summary for this PR Review Club is great and we could likely drop all of the Q&A section below, which I thought added less value.

Thanks, and I'm perfectly okay with that suggestion. In fact, I almost did that, but thought that _maybe_ keeping it would give the reader a feel for the level and kinds of detail and depth that occur in these release testing meetings -- I agree that the actual content is pretty low-information. But OTOH the reader can simply click the link to view the actual discussion log, so maybe that's not a good reason.

I just noticed, that there is no actual link to the review club page, I'll fix that right now (with a non-force push, new commit).

Feel free to remove the questions, or I can do that, or we can wait to see what others think. But I do like your suggestion.

Author: I'm +0.1 on keeping the QA.  I'm happy with whatever Larry and Mike decide.

---
title: 'Bitcoin Optech Newsletter #280'
permalink: /en/newsletters/2023/12/06/
name: 2023-12-06-newsletter
slug: 2023-12-06-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter describes several discussions about the proposed
cluster mempool and summarizes the results of a test performed using
warnet.  Also included are our regular sections summarizing a meeting of
the Bitcoin Core PR Review Club, announcing new releases and release
candidates, and describing notable changes to popular Bitcoin
infrastructure software.

## News

- **Cluster mempool discussion:** Bitcoin Core developers working on
  [cluster mempool][topic cluster mempool] started a [working
  group][wg-cluster-mempool] (WG) on the Delving Bitcoin forum. Cluster
  mempool is a proposal to make it much easier to operate on the mempool
  while respecting the required ordering of transactions.  A valid

>   ordering of Bitcoin transactions requires parent transactions be

Reviewer 5: ```suggestion
  ordering of Bitcoin transactions requires parent transactions to be
```

  confirmed before their children, either by putting the parent in an
  earlier block than any of its children, or earlier in the same block.
  In the cluster mempool design, _clusters_ of one or more related

>   transactions are designed to be to split into feerate-sorted _chunks_.

Reviewer 5: ```suggestion
  transactions are designed to be split into feerate-sorted _chunks_.
```

      As one of the theorems in the previously linked thread notes, "if
      one has two incomparable chunkings for a graph, then another
      chunking must exist which is strictly better than both".  That
      means an effective method for merging two different incomparable
      chunkings can be a powerful tool for improving miner
      profitability.  For example, a new transaction has been received that
      is related to other transactions already in the mempool, so its
      cluster needs to be updated, which implies its chunking also
      needs to be updated.  Two different methods of making that update
      can both be performed:

      1. A new chunking for the updated cluster is computed from
         scratch.  For large clusters, it may be computationally
         impractical to find an optimal chunking, so the new chunking
         might be less optimal than the old chunking.

      2. The previous chunking for the previous cluster is updated by
         inserting the new transaction in a location that is valid
         (parents before children).  This has the advantage that it
         preserves any existing optimizations in the unmodified chunks

>          but the downside that it could place the transaction in a

Reviewer 5: ```suggestion
         but the downside is that it could place the transaction in a
```

         suboptimal location.

      After the two different types of updates have been made, a

>       comparison may reveal one of them is strictly better, in which

Reviewer 5: ```suggestion
      comparison may reveal one of them to be strictly better, in which
```
or maybe "... may reveal that one of them is strictly better..."

      more fees than the original for any number of vbytes, and if it
      increases the total amount of fees in the mempool by enough to pay
      for its relay fees, then it should be included in the mempool.

        This evidence-based evaluation can replace several
        [heuristics][mempool replacements] currently used in Bitcoin
        Core to determine whether a transaction should be replaced,
        potentially improving the RBF rules in several areas, including
        proposed [package relay][topic package relay] for replacements.
        Several [other][cluster rbf-old1] threads [also][cluster
        rbf-old2] discussed [this][cluster rbf-old3] topic.

- **Testing with warnet:** Matthew Zipkin [posted][zipkin warnet] to
  Delving Bitcoin with the results of some simulations he's run using
  [warnet][], a program that launches a large number of Bitcoin nodes
  with a defined set of connections between them (usually on a test
  network).  Zipkin's results show the extra memory that would be used if
  several proposed changes to peer management code are merged (either
  independently or together).  He also notes that he's excited to use

>   simulations for testing other proposed changes and to quantify the

Reviewer 5: ```suggestion
  simulations for testing other proposed changes and quantifying the
```

It wastes developers' time to report bugs to them that aren't real bugs.
To mitigate these problems and promote testing efforts, a Review
Club meeting is held for a particular release candidate, in this instance, 26.0rc2.

The [26.0 release candidate testing guide][26.0 testing] was written by Max
Edwards, who also hosted the review club meeting with help from
Stéphan (stickies-v).

Attendees were also encouraged to get testing ideas by reading the
[26.0 release notes][].

This review club session covered two RPCs,
[`getprioritisedtransactions`][PR getprioritisedtransactions]
(also covered in an [earlier review club meeting][news250 pr review],
although the name of that RPC was changed after that review club meeting
was held), and [`importmempool`][PR importmempool].
The [New RPCs][] section of the release notes describes these and other
added RPCs.
The meeting also covered [V2 transport (BIP324)][topic v2 p2p transport],
and intended to cover [TapMiniscript][PR TapMiniscript] but this topic
wasn't discussed due to time limitations.

{% include functions/details-list.md
  q0="Which operating systems are people running?"

>   a0="Ubuntu 22.04 on WSL (Windows subsystem for Linux); macOs 13.4 (m1 chip)."

Reviewer 4: ```suggestion
  a0="Ubuntu 22.04 on WSL (Windows Subsystem for Linux); macOS 13.4 (M1 chip)."
```

  a2link="https://bitcoincore.reviews/v26-rc-testing#l-45"

  q3="What happens if we interrupt the CLI process during the import,
      then restart it (without stopping `bitcoind`)?"
  a3="This did not seem to cause any problem; the second import request
      completed as expected. It appears that the import process
      continued even after the CLI command was interrupted and
      that the second request didn't (for example) cause two import
      threads to run simultaneously and conflict with each other."
  a3link="https://bitcoincore.reviews/v26-rc-testing#l-91"

  q4="What are your results running V2 transport?"
  a4="The attendees weren't able to connect to a known mainnet
      V2-enabled node; it didn't seem to accept the connection request.
      It was suggested that all of its inbound slots may have been in use.
      Therefore, no P2P testing could be done during the meeting."
  a4link="https://bitcoincore.reviews/v26-rc-testing#l-115"
%}

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

> - [Bitcoin Core 26.0][] is a the next major version of the predominant
>   full-node implementation.  The release includes experimental support

Reviewer 4: ```suggestion
- [Bitcoin Core 26.0][] is the next major version of the predominant
  full node implementation.  The release includes experimental support
```

  for the [version 2 transport protocol][topic v2 p2p transport],
  support for [taproot][topic taproot] in [miniscript][topic
  miniscript], new RPCs for working with states for [assumeUTXO][topic
  assumeutxo], and an experimental RPC for processing [packages][topic
  package relay] of transactions (which is not yet supported for relay),
  among numerous other improvements and bug fixes.

- [LND 0.17.3-beta.rc1][] is a release candidate that contains several
  bug fixes.

## Notable code and documentation changes

*Notable recent changes in [Bitcoin Core][bitcoin core repo], [Core
Lightning][core lightning repo], [Eclair][eclair repo], [LDK][ldk repo],
[LND][lnd repo], [libsecp256k1][libsecp256k1 repo], [Hardware Wallet
Interface (HWI)][hwi repo], [Rust Bitcoin][rust bitcoin repo], [BTCPay
Server][btcpay server repo], [BDK][bdk repo], [Bitcoin Improvement
Proposals (BIPs)][bips repo], [Lightning BOLTs][bolts repo], and
[Bitcoin Inquisition][bitcoin inquisition repo].*

- [Bitcoin Core #28848][] updates the `submitpackage` RPC to be more helpful when

>   any transaction fails. Instead of throwing a JSONRPCError with the first

Reviewer 4: ```suggestion
  any transaction fails. Instead of throwing a `JSONRPCError` with the first
```
